### PR TITLE
handle long param attributes

### LINKED ISSUE #4032: Param attributes stripped when functions have certain length
Create a file with this input:

```rust
fn a1(#[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa] a: u8) {}
fn b1(#[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa] bb: u8) {}
fn a2(#[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa] a: u8) {}
fn b2(#[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa] bb: u8) {}
```

Run `cargo fmt`. `a1` and `b1` will have their attributes stripped. Remove one letter from the line and it will not be stripped as seen in `a2` and `b2`.

```rust
fn a1(a: u8) {}
fn b1(bb: u8) {}
fn a2(
    #[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa] a: u8,
) {
}
fn b2(
    #[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa] bb: u8,
) {
}
```

### PR DESCRIPTION
**Pull request by calebcartwright**
Closes #4032

### CODE REVIEW

File: rustfmt-core/rustfmt-lib/tests/target/issue_4032.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+fn a1(
+    #[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa] a: u8,
```
Comment (topecongiro) [inline rustfmt-core/rustfmt-lib/tests/target/issue_4032.rs]: Ideally, if the total width exceeds the max width, then the attribute and the parameter should be put on different lines:

```rust
fn a1(
    #[aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa]
    a: u8,
) {
}
```

@calebcartwright It would be great if you can fix this in this PR, though I am happy to merge this PR as is as entirely removing the attribute is much worse!

Comment (calebcartwright) [reply rustfmt-core/rustfmt-lib/tests/target/issue_4032.rs]: > Ideally, if the total width exceeds the max width, then the attribute and the parameter should be put on different lines

Agreed. AFAICT, the attribute and param will be split as desired in most cases. Based on some testing with this, I believe the following circumstances have to all be true in order for them to not be split:

* only one attribute on the param
* neither the attribute nor the param individually exceed the max width
* param is named and not an empty infer/closure param
* the combination of the formatted attributed and the param pattern is under the max width, but the addition of the colon and param type causes the total line length to exceed max width

That makes me think it's a bit of an edge case, but happy to try to find a way to handle it!

Comment (topecongiro) [reply rustfmt-core/rustfmt-lib/tests/target/issue_4032.rs]: @calebcartwright Thank you for investigating the issue further! I agree that this is an edge case and not problematic enough to block this PR. As such, I will merge this PR as is. If you find it interesting and fun, you are more than welcome to send the follow-up PR to fix this corner case.

Comment (calebcartwright) [reply rustfmt-core/rustfmt-lib/tests/target/issue_4032.rs]: I've opened #4059 in case anyone wants to work on it and to make sure we don't lose track of it. I'll try to circle back to it eventually if no one else does, but need to finish some other issues (for rustfmt 2.x, etc.) first 👍
